<a href="https://colab.research.google.com/github/piyush1856/Business-Case-OLA-Ensemble-Learning/blob/main/Business_Case_OLA_Ensemble_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduction**

Driver attrition poses significant challenges to ride-hailing platforms like OLA, impacting operational efficiency and customer satisfaction. Predicting and mitigating driver turnover is crucial for maintaining a stable and efficient workforce. This case study focuses on leveraging ensemble learning techniques to develop a robust predictive model for driver attrition.

By employing methods such as bagging and boosting, the objective is to harness the power of advanced machine learning algorithms to achieve higher predictive accuracy. The dataset includes driver demographics, tenure information, and performance metrics, which will be analyzed through exploratory data analysis, preprocessing, and feature engineering.

This study aims to deliver actionable insights and a reliable predictive model that can support strategic decision-making for managing driver retention effectively.

# **Background**

The ride-hailing industry has revolutionized transportation, offering convenience and flexibility to millions of customers globally. However, driver attrition remains one of the most critical challenges for companies like OLA. High turnover rates lead to increased recruitment and training costs, reduced service reliability, and a compromised customer experience.

Driver attrition is often influenced by a combination of factors, including demographic characteristics, work performance, and job tenure. Understanding these factors and predicting attrition early is vital for implementing proactive strategies to retain drivers and ensure operational stability.

Traditional machine learning models often fall short in capturing complex patterns within data. Ensemble learning, which combines multiple models to improve predictive accuracy and robustness, offers a promising solution to this challenge. Techniques like bagging and boosting can enhance model performance by addressing issues such as overfitting, bias, and variance.

This case study aims to explore the application of ensemble learning to predict driver attrition accurately, offering insights that can help OLA retain its workforce and maintain a competitive edge in the industry.

# **Concepts Used**

To address the challenge of predicting driver attrition effectively, the following concepts and methodologies are employed in this case study:  

1. **Exploratory Data Analysis (EDA):**  
   EDA involves examining the dataset to uncover patterns, trends, and relationships among features. It also helps identify missing values, outliers, and potential data quality issues.  

2. **Data Preprocessing:**  
   - **KNN Imputation:** Used to handle missing values by imputing them based on the nearest neighbors.  
   - **Feature Engineering:** Creation of new, meaningful features or transformation of existing ones to improve model performance.  
   - **Class Imbalance Treatment:** Techniques like oversampling, undersampling, or Synthetic Minority Oversampling Technique (SMOTE) to balance the dataset.  
   - **Standardization:** Scaling features to have zero mean and unit variance to ensure uniformity, especially important for models sensitive to feature magnitudes.  
   - **Encoding:** Converting categorical variables into numerical format using techniques like one-hot encoding or label encoding.  

3. **Ensemble Learning:**  
   Ensemble learning methods are applied to enhance the predictive power of the model:  
   - **Bagging:** Techniques such as Random Forest aggregate predictions from multiple decision trees to reduce variance and improve robustness.  
   - **Boosting:** Algorithms like Gradient Boosting or AdaBoost iteratively focus on correcting errors made by previous models, reducing bias and improving accuracy.  

4. **Model Evaluation:**  
   Evaluation metrics such as precision, recall, F1-score, and ROC-AUC are used to assess model performance. Special emphasis is placed on precision and recall to ensure a balanced trade-off between false positives and false negatives.  

5. **Insights and Recommendations:**  
   Based on the model’s predictions and analysis of important features, actionable recommendations are provided to mitigate driver attrition.  


# **Dataset Explanation**

The `ola_driver.csv` dataset includes features describing driver attributes, performance, and tenure. Below is a detailed explanation of the columns:

1. **MMMM-YY:** Reporting month and year.  
2. **Driver_ID:** A unique identifier assigned to every driver.  
3. **Age:** Age of the driver.  
4. **Gender:** Driver's gender (Male: `0`, Female: `1`).  
5. **City:** City code representing the city where the driver operates.  
6. **Education_Level:** Education level of the driver, categorized as:  
   - `0`: 10+  
   - `1`: 12+  
   - `2`: Graduate  
7. **Income:** Average monthly income of the driver.  
8. **Date Of Joining:** The date when the driver joined Ola.  
9. **LastWorkingDate:** The most recent or final day the driver worked with Ola.  
10. **Joining Designation:** The designation of the driver at the time of joining Ola.  
11. **Grade:** A grade assigned to the driver at the reporting time, likely representing performance or other metrics.  
12. **Total Business Value:** The total monetary value (business) a driver contributes in a month. Negative values might indicate cancellations, refunds, or other financial adjustments.  
13. **Quarterly Rating:** The driver's rating on a quarterly basis, ranging from `1` to `5` (with `5` being the best).  

This dataset provides essential details to analyze and predict driver attrition using ensemble learning techniques.


# **Import Libraries**

In [1]:
# Data manipulation and analysis
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Handling dates
from datetime import datetime

# Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.impute import KNNImputer

# Ensemble learning methods
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

# Model evaluation
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, roc_curve

# Warnings
import warnings
warnings.filterwarnings('ignore')

# **Define Problem Statement and perform Exploratory Data Analysis**

## **Definition of Problem**

Ola is a prominent player in the ride-hailing industry, relying heavily on its driver base to ensure consistent operations and meet customer demand. However, driver attrition poses a significant challenge, leading to operational disruptions, increased recruitment costs, and potential customer dissatisfaction.

The objective of this case study is to build a predictive model to identify drivers who are at risk of leaving the platform. By leveraging historical driver data, the model aims to:

- Proactively address attrition concerns by identifying at-risk drivers.
- Enable strategic interventions, such as targeted incentives or training, to retain drivers.
- Maintain a consistent and reliable driver base, ensuring business continuity and customer satisfaction.

The ability to predict driver churn is critical for Ola to stay competitive in the dynamic ride-hailing market while minimizing costs associated with driver turnover.


## **Observations on Data**

In [2]:
# Load the CSV file from the URL
url = "https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/002/492/original/ola_driver_scaler.csv"
df = pd.read_csv(url)

In [3]:
# Displaying the first few rows of the dataset to check the structure
df.head()

Unnamed: 0    MMM-YY  Driver_ID   Age  Gender City  Education_Level  \
0           0  01/01/19          1  28.0     0.0  C23                2   
1           1  02/01/19          1  28.0     0.0  C23                2   
2           2  03/01/19          1  28.0     0.0  C23                2   
3           3  11/01/20          2  31.0     0.0   C7                2   
4           4  12/01/20          2  31.0     0.0   C7                2   

   Income Dateofjoining LastWorkingDate  Joining Designation  Grade  \
0   57387      24/12/18             NaN                    1      1   
1   57387      24/12/18             NaN                    1      1   
2   57387      24/12/18        03/11/19                    1      1   
3   67016      11/06/20             NaN                    2      2   
4   67016      11/06/20             NaN                    2      2   

   Total Business Value  Quarterly Rating  
0               2381060                 2  
1               -665480                 2  
2                     0                 2  
3                     0                 1  
4                     0                 1

In [4]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [5]:
df.sample(10)

MMM-YY  Driver_ID   Age  Gender City  Education_Level  Income  \
12833  09/01/19       1901  28.0     0.0  C11                2   96265   
7058   02/01/19       1055  33.0     0.0   C6                2   70772   
17727  03/01/20       2594  33.0     0.0  C21                1  131568   
13115  06/01/20       1949  31.0     1.0   C3                1   88909   
3380   09/01/19        504  33.0     1.0  C15                1   49653   
6573   01/01/19        985  27.0     0.0  C25                1   49031   
10279  03/01/19       1531  31.0     0.0  C25                2   46001   
5253   08/01/20        781  41.0     1.0  C23                2   82503   
6203   07/01/20        927  28.0     0.0  C29                2   61371   
16596  02/01/20       2455  25.0     0.0  C28                1   99691   

      Dateofjoining LastWorkingDate  Joining Designation  Grade  \
12833      07/11/19        20/09/19                    3      3   
7058       26/05/18        02/09/19                    2      2   
17727      23/07/15             NaN                    1      4   
13115      29/10/19             NaN                    3      3   
3380       08/06/15        09/09/19                    2      2   
6573       12/11/16             NaN                    1      1   
10279      21/01/17        23/03/19                    2      2   
5253       21/12/19        29/07/20                    3      3   
6203       05/04/20             NaN                    1      1   
16596      18/06/17             NaN                    1      4   

       Total Business Value  Quarterly Rating  
12833                     0                 1  
7058                      0                 1  
17727               2400130                 2  
13115                -70510                 1  
3380                      0                 1  
6573                 219730                 2  
10279                     0                 1  
5253                      0                 1  
6203                 840480                 2  
16596                539590                 1

In [6]:
# Shape of data
print("No. of rows: ", df.shape[0])
print("No. of columns: ", df.shape[1])

No. of rows:  19104
No. of columns:  13


In [7]:
# Overview
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19104 entries, 0 to 19103
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   MMM-YY                19104 non-null  object 
 1   Driver_ID             19104 non-null  int64  
 2   Age                   19043 non-null  float64
 3   Gender                19052 non-null  float64
 4   City                  19104 non-null  object 
 5   Education_Level       19104 non-null  int64  
 6   Income                19104 non-null  int64  
 7   Dateofjoining         19104 non-null  object 
 8   LastWorkingDate       1616 non-null   object 
 9   Joining Designation   19104 non-null  int64  
 10  Grade                 19104 non-null  int64  
 11  Total Business Value  19104 non-null  int64  
 12  Quarterly Rating      19104 non-null  int64  
dtypes: float64(2), int64(7), object(4)
memory usage: 1.9+ MB


In [8]:
# Statistical Summary
df.describe()

Driver_ID           Age        Gender  Education_Level  \
count  19104.000000  19043.000000  19052.000000     19104.000000   
mean    1415.591133     34.668435      0.418749         1.021671   
std      810.705321      6.257912      0.493367         0.800167   
min        1.000000     21.000000      0.000000         0.000000   
25%      710.000000     30.000000      0.000000         0.000000   
50%     1417.000000     34.000000      0.000000         1.000000   
75%     2137.000000     39.000000      1.000000         2.000000   
max     2788.000000     58.000000      1.000000         2.000000   

              Income  Joining Designation         Grade  Total Business Value  \
count   19104.000000         19104.000000  19104.000000          1.910400e+04   
mean    65652.025126             1.690536      2.252670          5.716621e+05   
std     30914.515344             0.836984      1.026512          1.128312e+06   
min     10747.000000             1.000000      1.000000         -6.000000e+06   
25%     42383.000000             1.000000      1.000000          0.000000e+00   
50%     60087.000000             1.000000      2.000000          2.500000e+05   
75%     83969.000000             2.000000      3.000000          6.997000e+05   
max    188418.000000             5.000000      5.000000          3.374772e+07   

       Quarterly Rating  
count      19104.000000  
mean           2.008899  
std            1.009832  
min            1.000000  
25%            1.000000  
50%            2.000000  
75%            3.000000  
max            4.000000

### **Statistical Summary Inference**

- **Age**: The average age of drivers is around 34.67 years, with a minimum of 21 years and a maximum of 58 years.
- **Gender**: The dataset shows a higher number of male drivers (mean value of 0.42 for gender).
- **Education Level**: The majority of drivers have completed their schooling (Education Level = 0 or 1).
- **Income**: The average monthly income of drivers is approximately 65,652 with a wide range (from 10,747 to 188,418).
- **Total Business Value**: There's significant variation in drivers' total business value, with some negative values, indicating possible refunds or cancellations.
- **Quarterly Rating**: The most frequent rating is 2, with a minimum of 1 and maximum of 4, reflecting the variance in driver performance.


## **Datatype Conversion**

In [9]:
## Converting 'MMM-YY' feature to datetime type
df['MMM-YY'] = pd.to_datetime(df['MMM-YY'])

## Converting 'Dateofjoining' feature to datetime type
df['Dateofjoining'] = pd.to_datetime(df['Dateofjoining'])

## Converting 'LastWorkingDate' feature to datetime type
df['LastWorkingDate'] = pd.to_datetime(df['LastWorkingDate'])

In [10]:
# Converting to categorical columns
categorical_columns = ['Gender', 'Education_Level', 'City']
for col in categorical_columns:
    df[col] = df[col].astype('category')

In [11]:
df.dtypes

MMM-YY                  datetime64[ns]
Driver_ID                        int64
Age                            float64
Gender                        category
City                          category
Education_Level               category
Income                           int64
Dateofjoining           datetime64[ns]
LastWorkingDate         datetime64[ns]
Joining Designation              int64
Grade                            int64
Total Business Value             int64
Quarterly Rating                 int64
dtype: object

## **Missing Value Detection and Treatment**

In [12]:
# Getting count of null/missing values in dataset
null_values = df.isnull().sum().reset_index(name='null_count')
null_values.sort_values(by='null_count',ascending=False,inplace=True)
null_values.reset_index(drop=True,inplace=True)
null_values

index  null_count
0        LastWorkingDate       17488
1                    Age          61
2                 Gender          52
3                 MMM-YY           0
4              Driver_ID           0
5                   City           0
6        Education_Level           0
7                 Income           0
8          Dateofjoining           0
9    Joining Designation           0
10                 Grade           0
11  Total Business Value           0
12      Quarterly Rating           0

In [13]:
# Ensure 'Gender' is numeric
df['Gender'] = df['Gender'].astype('category').cat.codes

In [14]:
# Select only numeric columns and exclude 'Driver_ID'
numeric_columns = df.select_dtypes(include=['float64', 'int64']).columns
columns_for_imputation = numeric_columns.drop('Driver_ID')
columns_for_imputation

Index(['Age', 'Income', 'Joining Designation', 'Grade', 'Total Business Value',
       'Quarterly Rating'],
      dtype='object')

In [15]:
# Initialize the KNN imputer
knn_imputer = KNNImputer(n_neighbors=5, weights='uniform', metric='nan_euclidean')

In [16]:
# Apply the imputer to the selected columns
imputed_data = knn_imputer.fit_transform(df[columns_for_imputation])

In [17]:
# Replace the imputed values back into the DataFrame
df[columns_for_imputation] = imputed_data

In [18]:
# Convert 'Gender' back to categorical
df['Gender'] = df['Gender'].astype('category')

In [19]:
# Getting count of null/missing values in dataset
null_values = df.isnull().sum().reset_index(name='null_count')
null_values.sort_values(by='null_count',ascending=False,inplace=True)
null_values.reset_index(drop=True,inplace=True)
null_values

index  null_count
0        LastWorkingDate       17488
1                 MMM-YY           0
2              Driver_ID           0
3                    Age           0
4                 Gender           0
5                   City           0
6        Education_Level           0
7                 Income           0
8          Dateofjoining           0
9    Joining Designation           0
10                 Grade           0
11  Total Business Value           0
12      Quarterly Rating           0

In the context of the dataset, the `LastWorkingDate` column represents the most recent or final day a driver worked with Ola. It is important to note that the `LastWorkingDate` column will only contain missing values (`NaN`) for drivers who are still active, as they haven't yet had a last working day recorded.

## **Target Variable Creation**

In [20]:
# Creating Target Variable
df['Is_Churned'] = df['LastWorkingDate'].notnull().astype('category')

In [21]:
df.head()

MMM-YY  Driver_ID   Age Gender City Education_Level   Income  \
0 2019-01-01          1  28.0      0  C23               2  57387.0   
1 2019-02-01          1  28.0      0  C23               2  57387.0   
2 2019-03-01          1  28.0      0  C23               2  57387.0   
3 2020-11-01          2  31.0      0   C7               2  67016.0   
4 2020-12-01          2  31.0      0   C7               2  67016.0   

  Dateofjoining LastWorkingDate  Joining Designation  Grade  \
0    2018-12-24             NaT                  1.0    1.0   
1    2018-12-24             NaT                  1.0    1.0   
2    2018-12-24      2019-03-11                  1.0    1.0   
3    2020-11-06             NaT                  2.0    2.0   
4    2020-11-06             NaT                  2.0    2.0   

   Total Business Value  Quarterly Rating Is_Churned  
0             2381060.0               2.0      False  
1             -665480.0               2.0      False  
2                   0.0               2.0       True  
3                   0.0               1.0      False  
4                   0.0               1.0      False

## **Creation of Aggregated Datasets**

In [22]:
aggregation_dict = {
    'Age': 'max',                     # Use the maximum age in the period (driver's age is unlikely to change)
    'Gender': 'first',                # Gender remains constant, so first value in the group is used
    'City': 'first',                  # City is generally constant, so the first recorded city is used
    'Education_Level': 'last',        # Education level is typically unchanged or select latest, use the last value
    'Income': 'last',                 # Retain the most recent income value for the driver
    'Dateofjoining': 'last',          # Date of joining is fixed for each driver, but we'll take the most recent value in case of multiple records
    'LastWorkingDate': 'last',        # Keep the most recent working date, NaT will be retained for inactive periods
    'Joining Designation': 'last',    # Designation generally doesn't change often, use the most recent designation
    'Grade': 'last',                  # Grade can change, so we take the most recent grade
    'Total Business Value': 'sum',    # Summing the total business value across periods
    'Quarterly Rating': 'last',       # The last quarterly rating will give the most up-to-date rating
    'Is_Churned': 'last'              # The churn flag will be taken from the last record (if a driver churned, this will be 1)
}

# Apply grouping based on Driver_ID and MMM-YY, then aggregate according to the aggregation_dict
df_agg_driver_id_and_reporting = df.groupby(['Driver_ID', 'MMM-YY']).aggregate(aggregation_dict)

# Sort the data by Driver_ID and MMM-YY
df_agg_driver_id_and_reporting = df_agg_driver_id_and_reporting.sort_index(ascending=[True,True])
df_agg_driver_id_and_reporting.head()

Age Gender City Education_Level   Income Dateofjoining  \
Driver_ID MMM-YY                                                                
1         2019-01-01  28.0      0  C23               2  57387.0    2018-12-24   
          2019-02-01  28.0      0  C23               2  57387.0    2018-12-24   
          2019-03-01  28.0      0  C23               2  57387.0    2018-12-24   
2         2020-11-01  31.0      0   C7               2  67016.0    2020-11-06   
          2020-12-01  31.0      0   C7               2  67016.0    2020-11-06   

                     LastWorkingDate  Joining Designation  Grade  \
Driver_ID MMM-YY                                                   
1         2019-01-01             NaT                  1.0    1.0   
          2019-02-01             NaT                  1.0    1.0   
          2019-03-01      2019-03-11                  1.0    1.0   
2         2020-11-01             NaT                  2.0    2.0   
          2020-12-01             NaT                  2.0    2.0   

                      Total Business Value  Quarterly Rating Is_Churned  
Driver_ID MMM-YY                                                         
1         2019-01-01             2381060.0               2.0      False  
          2019-02-01             -665480.0               2.0      False  
          2019-03-01                   0.0               2.0       True  
2         2020-11-01                   0.0               1.0      False  
          2020-12-01                   0.0               1.0      False

In [23]:
# Aggregation dictionary for further aggregation on the already grouped data
function_dict = {
    'Age': 'mean',                    # Average age of the driver over all months
    'Gender': 'first',                # First gender encountered (as gender is unlikely to change)
    'City': 'first',                  # First city recorded for the driver
    'Education_Level': 'last',        # Most recent education level
    'Income': 'mean',                 # Average income over time
    'Joining Designation': 'last',    # Most recent joining designation
    'Grade': 'last',                  # Most recent grade achieved
    'Total Business Value': 'sum',    # Total business value over all months
    'Quarterly Rating': 'mean',       # Average quarterly rating over time
    'Is_Churned': 'last',             # Churn status (the last value will indicate if the driver churned)
    'Dateofjoining': 'last',          # Most recent joining date
    'LastWorkingDate': 'last'         # Most recent working date (NaT if no working)
}

# Apply secondary aggregation on the already grouped data
final_aggregated_data = df_agg_driver_id_and_reporting.groupby('Driver_ID').agg(function_dict)

In [24]:
# Drop 'Driver_ID' and reset the index
final_aggregated_data = final_aggregated_data.reset_index(drop=True)

In [25]:
final_aggregated_data.head()

Age Gender City Education_Level   Income  Joining Designation  Grade  \
0  28.0      0  C23               2  57387.0                  1.0    1.0   
1  31.0      0   C7               2  67016.0                  2.0    2.0   
2  43.0      0  C13               2  65603.0                  2.0    2.0   
3  29.0      0   C9               0  46368.0                  1.0    1.0   
4  31.0      1  C11               1  78728.0                  3.0    3.0   

   Total Business Value  Quarterly Rating Is_Churned Dateofjoining  \
0             1715580.0               2.0       True    2018-12-24   
1                   0.0               1.0      False    2020-11-06   
2              350000.0               1.0       True    2019-12-07   
3              120360.0               1.0       True    2019-01-09   
4             1265000.0               1.6      False    2020-07-31   

  LastWorkingDate  
0      2019-03-11  
1             NaT  
2      2020-04-27  
3      2019-03-07  
4             NaT